src/cpp/session/modules/SessionThemes.R

```diff
@@ -0,0 +1,977 @@
+      {
+         stop(
+            "expected 3 values for RGB color, not ",
+            length(color),
+            call. = FALSE)
+      }
+      colorVec <- color
+   }
+   else if (substr(color, 0, 1) == "#") 
+   {
+      if (nchar(color) != 7)
+      {
+         stop(
+            "hex represntation of RGB values should have the format \"#RRGGBB\", where `RR`, `GG` and `BB` are in [0x00, 0xFF]. Found: ",
+            color,
+            call. = FALSE)
+      }
+      else
+      {
+         colorVec <- sapply(
+            c(substr(color, 2, 3), substr(color, 4, 5), substr(color, 6, 7)),
+            hexStrToI <- function(str) { strtoi(str, 16L) },
+            USE.NAMES = FALSE)
+      }
+   }
+   else if (grepl("^rgba?", color))
+   {
+      matches = regmatches(color, regexec("\\(([^,\\)]+),([^,\\)]+),([^,\\)]+)", color))[[1]]
+      if (length(matches) != 4)
+      {
+         stop(
+            "non-hex representation of RGB values should have the format \"rgb(R, G, B)\" or \"rgba(R, G, B, A)\" where `R`, `G`, and `B` are integer values in [0, 255] and `A` is decimal value in [0, 1.0]. Found: ",
+            color,
+            call. = FALSE)
+      }
+      colorVec <- strtoi(matches[2:4])
+   }
+   else
+   {
+      stop(
+         "supplied color has an invalid format: ",
+         color,
+         ". Expected \"#RRGGBB\", \"rgb(R, G, B) or \"rgba(R, G, B, A)\", where `RR`, `GG` and `BB` are in [0x00, 0xFF], `R`, `G`, and `B` are integer values in [0, 255], and `A` is decimal value in [0, 1.0]",
+         call. = FALSE)
+   }
+   
+   # Check for inconsistencies.
+   invalidMsg <- paste0("invalid color supplied: ", color, ". ")
+   if (any(is.na(colorVec)) || any(!is.integer(colorVec)))
+   {
+      stop(
+         invalidMsg,
+         "One or more RGB values could not be converted to an integer",
+         call. = FALSE)
+   }
+   if (any(colorVec < 0))
+   {
+      stop(invalidMsg, "RGB value cannot be negative", call. = FALSE)
+   }
+   if (any(colorVec > 255))
+   {
+      stop(invalidMsg, "RGB value cannot be greater than 255", call. = FALSE)
+   }
+   
+   colorVec
+})
+
+# Mixes two colours together.
+#
+# @param color1   The first colour.
+# @param color2   The second colour.
+# @param alpha1   The alpha of the first colour.
+# @param alpha2   The alpha of the second colour.
+# 
+# Returns the mixed colour in string format.
+.rs.addFunction("mixColors", mixColors <- function(color1, color2, alpha1, alpha2 = NULL) {
+   c1rgb <- .rs.getRgbColor(color1)
+   c2rgb <- .rs.getRgbColor(color2)
+   
+   if (is.null(alpha2))
+   {
+      alpha2 = 1 - alpha1
+   }
+   
+   paste0(
+      "rgb(",
+      paste(
+         round(alpha1 * c1rgb[[1]] + alpha2 * c2rgb[[1]]),
+         round(alpha1 * c1rgb[[2]] + alpha2 * c2rgb[[2]]),
+         round(alpha1 * c1rgb[[3]] + alpha2 * c2rgb[[3]]),
+         sep = ","),
+      ")")
+})
+
+# Determines the luma of a colour. This refers to the percieved luminance of the colour. More 
```

Comment:
Now I'm being super-annoying. Mainly because I haven't written a lot of R so not great at reviewing it so end up focusing on nit-picking. Typo in "perceived".